### PR TITLE
Support matching resources in Java try-with-resources

### LINKED ISSUE #717: Java: semgrep does not match the 'try-with-resources' pattern.
**Describe the bug**
The 'try-with-resources' pattern was added in Java 8: https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

This example I was using as a test case uses this pattern, but semgrep doesn't pick it up:
https://find-sec-bugs.github.io/bugs.htm#OBJECT_DESERIALIZATION

**To Reproduce**
This pattern should match both cases:
https://semgrep.live/GD1

It looks like it just gets skipped in the AST altogether:
image

### PR DESCRIPTION
**Pull request by emjin**
Closes #717, #4228

Translate try with resources to WithUsingResource and Try. This allows people to find problems in how the resources are acquired.

Test plan: make test

PR checklist:
- [x] Documentation is up-to-date
- [x] Changelog is up-to-date
- [x] Change has no security implications (otherwise, ping security team)

### CODE REVIEW

File: semgrep-core/src/parsing/pfff/java_to_generic.ml
Comment (IagoAbal) [inline semgrep-core/src/parsing/pfff/java_to_generic.ml]: I think you could merge the cases `[]` and `_ :: _`. Do we need a special case for `[ v ]` ?

File: semgrep-core/src/parsing/pfff/java_to_generic.ml
Comment (IagoAbal) [inline semgrep-core/src/parsing/pfff/java_to_generic.ml]: Can easily avoid `unsafe_fake` here by passing a token to `resource`, which we have readily available in `resources`.

File: semgrep-core/src/parsing/pfff/java_to_generic.ml
Comment (emjin) [reply semgrep-core/src/parsing/pfff/java_to_generic.ml]: I wasn't sure if it would be better to avoid the unnecessary Block, can take it out

**Review by IagoAbal — APPROVED** [review]
Nice! Thanks for taking care of this @emjin! I would appreciate if you get rid of the `unsafe_fake` before merging, we should avoid unsafe fake tokens unless it's really a hassle to use safe ones.

File: semgrep-core/src/parsing/pfff/java_to_generic.ml
Comment (IagoAbal) [reply semgrep-core/src/parsing/pfff/java_to_generic.ml]: I would keep the translation trivial unless there is a good reason to save that `Block`. If you prefer to save the `Block` I would merge `[]` and `_ :: _` since they are the same. But these are just minor observations anyways. It's your call!

File: semgrep-core/src/parsing/pfff/java_to_generic.ml
Comment (nbrahms) [inline semgrep-core/src/parsing/pfff/java_to_generic.ml]: What happens to this code:
```
    try (
        java.util.zip.ZipFile zf =
             new java.util.zip.ZipFile(zipFileName);
        java.io.BufferedWriter writer = 
            java.nio.file.Files.newBufferedWriter(outputFilePath, charset)
    ) {
        // Enumerate each entry
        for (java.util.Enumeration entries =
                                zf.entries(); entries.hasMoreElements();) {
            // Get the entry name and write it to the output file
            String newLine = System.getProperty("line.separator");
            String zipEntryName =
                 ((java.util.zip.ZipEntry)entries.nextElement()).getName() +
                 newLine;
            writer.write(zipEntryName, 0, zipEntryName.length());
        }
    }
```
(found on the Java documentation at https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html)

Comment (nbrahms) [reply semgrep-core/src/parsing/pfff/java_to_generic.ml]: E.g., can we have a test that matches against multiple resources?

Comment (emjin) [reply semgrep-core/src/parsing/pfff/java_to_generic.ml]: It's a block of resources (the semicolon token is lost, may be a problem one day). I added a test to cover it

**Review by nbrahms — APPROVED** [review]
lgtm

**Comment by emjin** [comment]
Error because of https://github.com/returntocorp/semgrep-rules/pull/1763, I forget what the preferred way to resolve the cycle is

Edit: answer is merge PR with errors to develop branch of semgrep-rules

File: CHANGELOG.md
Comment (IagoAbal) [inline CHANGELOG.md:22]: Just spotted this when I was going to do the release, but note that you added the entry in the wrong place.

**Pull request merged**
